Improve messaging on volume expansion

gnufied opened the pull request:
- we now provide clear message to user what to do when cloudprovider resizing is finished
  and file system resizing is needed.
- add a event when resizing is successful
- Use PATCH both in controller-manager and kubelet for updating PVC status
- Remove code duplication between controller-manager and kubelet for updating PVC status
- Only remove conditions that are managed by resize controller



```release-note
Improve messages user gets during and after volume resizing is done.
```

gnufied committed:
Improve messaging on resize

- we now provide clear message to user what to do when cloudprovider resizing is finished
  and file system resizing is needed.
- add a event when resizing is successful.
- Use Patch for updating PVCs in both kubelet and controller-manager
- Extract updating pvc util function in one place.
- Only update resize conditions on progress

gnufied commented:
/sig storage

cc @jsafrane @saad-ali

gnufied commented:
/test pull-kubernetes-unit

jsafrane commented:
/assign

* What removes the condition? I see it's only set and never removed / set to False.
* What is the UX? Will `kubect describe pvc` show the condition?
* An e2e test update would help :-)

gnufied commented:
The condition is removed here - https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/operationexecutor/operation_generator.go#L1372

Currently we don't seem to be showing Conditions in pvc describe. Will fix this in a follow up PR. 

e2e test.. okay, will add. :-)

jsafrane commented on pkg/controller/volume/expand/cache/volume_resize_map.go:
```diff
@@ -160,6 +163,19 @@ func (resizeMap *volumeResizeMap) MarkAsResized(pvcr *PVCWithResizeRequest, newS
+		Type:               v1.PersistentVolumeClaimFileSystemResizePending,
+		Status:             v1.ConditionTrue,
+		LastTransitionTime: metav1.Now(),
+		Message:            "Restart pod that is using this claim to finish File system resize on node",
```
IMO, it should say why the condition is set. "Waiting for user to (re-)start a pod to finish resize of the volume" or something in this direction?

gnufied replied:
fixed

jsafrane commented:
> The condition is removed here 

Ack

> Currently we don't seem to be showing Conditions in pvc describe. Will fix this in a follow up PR.

Ack

jsafrane commented:
/lgtm
for the controller changes Need API review/approvals.

gnufied commented:
Flaked on https://github.com/kubernetes/kubernetes/issues/58387

/test pull-kubernetes-e2e-kops-aws

gnufied commented:
/test pull-kubernetes-e2e-kops-aws

liggitt commented:
> The condition is removed here - https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/operationexecutor/operation_generator.go#L1372

We *cannot* blindly remove/reset all conditions. Anywhere this flow is doing that needs to be corrected. I saw at least these places:

https://github.com/kubernetes/kubernetes/blob/6ec4cb107e0a2d698d107ce2ee482173966a16b6/pkg/volume/util/operationexecutor/operation_generator.go#L1372

https://github.com/kubernetes/kubernetes/blob/6ec4cb107e0a2d698d107ce2ee482173966a16b6/pkg/controller/volume/expand/cache/volume_resize_map.go#L153-L155

updatePVCCapacityAndConditions should be modified so that it doesn't encourage wholesale replacement of status conditions

1. Only add/remove/modify conditions you explicitly know about. That maintains interoperability with other components adding conditions for their own purposes.
2. If you are using conditions to coordinate between multiple components, be aware of the maximum version skew between those components (a controller setting a new condition and expecting it to be updated by a kubelet has to consider that kubelets up to 2 releases old will be processing this object... in this case, it might be accurate for the condition to continue reflecting a pending resize if an old kubelet wasn't paying any attention to the resize). Skew is less of an issue for features when they're in alpha. The point at which a feature goes beta, skew has to be considered carefully.

The kubelet/node condition updating code is a good example to look at:

https://github.com/kubernetes/kubernetes/blob/6ec4cb107e0a2d698d107ce2ee482173966a16b6/pkg/kubelet/kubelet_node_status.go#L805-L823

It updates a specific condition, leaving all others undisturbed.

gnufied commented:
@liggitt I have updated the code, so as it only considers Conditions that volume expand controller explicitly  knows about. Also, I have updated PVC update to use PATCH in both controller-manager and kubelet while I was at it. PTAL

gnufied commented:
/retest

gnufied commented:
@jsafrane this needs another round of lgtm after taking into account @liggitt feedback.

jsafrane commented on pkg/controller/volume/expand/cache/volume_resize_map.go:
```diff
@@ -195,15 +211,13 @@ func (resizeMap *volumeResizeMap) UpdatePVSize(pvcr *PVCWithResizeRequest, newSi
+		pvcr.PVC,
+		pvcConditions,
+		resizeMap.kubeClient,
+		func(newPVC *v1.PersistentVolumeClaim) *v1.PersistentVolumeClaim {
```
Why indirection? It makes the code hard to read. IMO, it would be more readable to have something like:

```go
newPVC := pvcr.PVC.DeepCopy()
newPVC.Status.Capacity[v1.ResourceStorage] = newSize
UpdateCondition(newPVC, conditionName, newValue)
err := PatchPVC(resizeMap.kubeClient, pvcr.PVC /*old PVC*/, newPVC)
```

We could use PatchPVC everywhere we update PVCs

gnufied replied:
fixed.

jsafrane commented on pkg/volume/util/resize_util.go:
```diff
@@ -0,0 +1,98 @@
+// leaving other conditions untouched.
+func MergeResizeConditionOnPVC(
+	pvc *v1.PersistentVolumeClaim,
+	pvcConditions []v1.PersistentVolumeClaimCondition) *v1.PersistentVolumeClaim {
```
`pvcConditions` is quite confusing name, I stared at the code for a while to find out what's going on. Perhaps `resizeConditions`?

gnufied replied:
renamed

jsafrane commented:
/lgtm

gnufied commented:
/test pull-kubernetes-e2e-kops-aws

gnufied commented:
/test pull-kubernetes-e2e-kops-aws

gnufied commented:
AWS tests are flaking on  #58387 , unreleated to this PR

caesarxuchao commented:
@saad-ali for triage.

/unassign @caesarxuchao

liggitt commented on pkg/controller/volume/expand/cache/volume_resize_map.go:
```diff
@@ -160,6 +163,21 @@ func (resizeMap *volumeResizeMap) MarkAsResized(pvcr *PVCWithResizeRequest, newS
+	}
+	conditions := []v1.PersistentVolumeClaimCondition{pvcCondition}
+	newPVC := pvcr.PVC.DeepCopy()
+	newPVC = util.MergeResizeConditionOnPVC(newPVC, conditions)
```
take a look at what other condition updaters do... if the PersistentVolumeClaimFileSystemResizePending condition already exists with the status ConditionTrue, I wouldn't expect LastTransitionTime to be updated on the object. as written, MergeResizeConditionOnPVC doesn't consider whether the condition has actually updated, and always results in a write

gnufied replied:
I am not sure - that is true.

If `PersistentVolumeClaimFileSystemResizePending` already exists on the PVC and user has requested another resize while resize was pending on the node, the code will try and resize the PVC again and set a new `PersistentVolumeClaimFileSystemResizePending` with a new `LastTransitionTime` and it would be incorrect to not update it..

liggitt replied:
> If PersistentVolumeClaimFileSystemResizePending already exists on the PVC and user has requested another resize while resize was pending on the node, the code will try and resize the PVC again and set a new PersistentVolumeClaimFileSystemResizePending with a new LastTransitionTime and it would be incorrect to not update it..

I don't think that's accurate... in that scenario, the condition status has not changed, so the last transition time should not change

gnufied replied:
fixed

gnufied replied:
I have fixed this as per ^ discussion. But I am curious - why  does it matter? What happens if same condition (with same status) does end up changing `LastTransitionTime`?

liggitt replied:
that tends to lead to what should be a no-op API write continuously generating changes/writes/watch events

gnufied replied:
ack. can you please approve the PR?

gnufied commented:
/test pull-kubernetes-verify

gnufied commented:
@liggitt I believe I have addressed all your concerns. I ran e2e test for an hour in a tight loop and it looked good. Can you PTAL?

liggitt commented on pkg/volume/util/resize_util.go:
```diff
@@ -0,0 +1,126 @@
+	for _, condition := range oldConditions {
+		// If Condition is of not resize type, we keep it.
+		if _, ok := knownResizeConditions[condition.Type]; !ok {
+			newConditions = append(newConditions, condition)
```
this reorders conditions, which we don't want. see https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_node_status.go#L770-L788 for an example of in-place updating of an existing condition, then appending to the end if no existing condition is found

gnufied replied:
I fixed this. I did not quite use same approach as code you linked but the end result should be same.

liggitt commented:
once the condition reordering/updating is worked out, I think this looks right, though I'm having trouble following where each of the two conditions are added/removed (controller-side or kubelet-side). Will defer to the storage reviewer for those details.

liggitt commented on pkg/volume/util/resize_util_test.go:
```diff
@@ -0,0 +1,165 @@
+				{
+					Type:               v1.PersistentVolumeClaimResizing,
+					Status:             v1.ConditionTrue,
+					LastTransitionTime: metav1.Now(),
```
add time to this so it is distinctly different than currentTime

gnufied replied:
fixed

liggitt commented:
one nit on test, condition handling looks good to me otherwise. will defer to storage reviewer for lgtm on overall flow.
/approve

gnufied commented:
@jsafrane can you PTAL.

gnufied commented:
@liggitt fixed the minor nit, you pointed out in test case. 

/assign @derekwaynecarr

childsb commented:
/lgtm

gnufied commented:
ping @derekwaynecarr @smarterclayton , it would be really nice to have this approved. The pending approval in `pkg/kubelet` is really minor and just adds an event.

derekwaynecarr commented:
kubelet changes lgtm.

/approve

gnufied commented:
/test pull-kubernetes-e2e-kops-aws
